### ISSUE TITLE
flb_service_set() weird behaviour

### ISSUE THREAD

**Issue by valande** [fluent/fluent-bit#5359]
Hi all.

I apologyze in advance if this is not the correct way for reporting a probable issue, it is my first time doing that.

Trying to configure fluentbit service using _flb_service_set()_ api call, as far as I can see, it works pretty well if passed parameters are literals, but always returns -1 (error) if I use variables as parameters instead of constant literals.

I tried using _char *, const char *, std::string.data(), std::string.c_str(),_ etc, etc. and always getting the same return value -1.

I am stucked in this point.
Please, I would appreciate your comments if anyone here knows how to workaround this. 

Example that works fine:
`flb_service_set (m_ctx, "Flush", "20", "Log_level", "trace", "Grace", "10", "Parsers_File", "parsers.conf", NULL);`

Example that never worked:
(serviceParameters is of type _**std::unordered_map<std::string, const char *>**_)
`flb_service_set (m_ctx, "Flush",              serviceParameters["flush"], 
                                       "Log_level",       serviceParameters["loglevel"], 
                                       "Grace",            serviceParameters["grace"], 
                                       "Parsers_File",  serviceParameters["parsers_file"],
                                       nullptr);
`

Thanks in advance, and kind regards

**Comment by patrick-stephens** [comment]
It would help to know how it doesn't work, is there an error at compile time, etc.? What platform are you compiling for? Please follow the template so we capture this info.

The parameters are just a standard variable list: https://c-for-dummies.com/blog/?p=3398
Looking at the handling code here they're just standard `char *` pointers: https://github.com/fluent/fluent-bit/blob/master/src/flb_lib.c#L546-L559. I'm guessing some conversion funny business in the compiler as you're missing types above between literals and elements. It can probably be simplified down to a simple C reproducer using variable lists to verify exactly what is going wrong - I don't think there is any inherent issue with the actual library call.

Note that variable lists have their own syntax including things like passing the side of the list first (in `m_ctx`) but this is not Fluent Bit specific.

**Comment by valande** [comment]
Hi. Thanks for your reply. 

May I did not explain so good as I wanted, sorry. I finally got this problem solved, not as I expected, but could work-around it.

I said _"weird behaviour"_ because when I had a look to the source code I saw (almost) the same implementation for the methods _"flb_service_set()"_ and _"flb_output_set()"_. On this basis, knowing both of them use a standard variable list, as you noted above, I thought they would do similar work, but they finally do not.

When calling _"flb_output_set()"_ I can mix the configuration of _[OUTPUT]_ section, first reading a .conf file (kind of default or master configuration), later modifying some of the parameters over what was read. This worked smooth from the first time.

It is not the same for "flb_service_set()", always getting an error return (-1) when applying the same steps as worked for _"flb_output_set()"_. 
So, I finally decided to completely strip the _[SERVICE]_ section from the config file and setting the service parameters using _"flb_service_set()"_ call from source code only. And it is working well till now.

Regards,
Pablo.
